### PR TITLE
Move validators and OAuth outside of client

### PR DESCRIPTION
**Pull request by bakura10**
Ping @danizord

This backport a part of my change in #4 to the current version. This will require to update all our code to inject RequestValidator / WebhookValidator, and being able to retrieve the app metadata when necessary to inject shared secret (which is hence not needed anymore in the client).

Please review :D.

PS: I've introduced a dependency to Guzzle 6 to have access to the latest client, so that when we update to Guzzle 6 we do not need to change that part.

### CODE REVIEW

File: src/Container/ShopifyClientFactory.php [hunk 1]
```diff
@@ -51,18 +50,6 @@ public function __invoke(ContainerInterface $container)
...
-                throw new RuntimeException('Options "shop", "api_key" and "password" are mandatory when creating Shopify client for a private app');
-            }
-        } else {
-            if (!isset($config['api_key']) || !isset($config['shared_secret'])) {
-                throw new RuntimeException('Options "api_key" and "shared_secret" are mandatory when creating Shopify client for a public app');
-            }
-        }
-
-        return new ShopifyClient($config);
+        return new ShopifyClient($config['zfr_shopify']);
```
Comment (danizord) [inline src/Container/ShopifyClientFactory.php]: Should throw exception if `zfr_shopify` is undefined index

Comment (danizord) [inline src/Container/ShopifyClientFactory.php]: nvm :D

File: src/OAuth/Authorizer.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+     *
+     * @param  string $apiKey
+     * @param  string $shopDomain
+     * @param  array  $scopes
+     * @param  string $redirectionUri
+     * @param  string $nonce
+     * @return ResponseInterface
+     * @throws Exception\MissingApiKeyException
+     */
+    public function createAuthorizationResponse(string $apiKey, string $shopDomain, array $scopes, string $redirectionUri, string $nonce)
```
Comment (danizord) [inline src/OAuth/Authorizer.php]: Can be moved to it's own class extending Diactoros' RedirectResponse

File: src/Validator/RequestValidator.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * According to Shopify, a shop hostname must ends by "myshopify.com", and must only contains
+     * letters, numbers, dots and hyphens
+     *
+     * @param  array $queryParams
+     * @return void
+     * @throws Exception\InvalidRequestException
+     */
+    private function validateShopHostname(array $queryParams)
```
Comment (danizord) [inline src/Validator/RequestValidator.php]: Can be moved to `ShopDomain` value object

Comment (bakura10) [inline src/Validator/RequestValidator.php]: What would be the benefit? We need to have the complete query params anyway to validate the request.

File: src/Validator/WebhookValidator.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+     */
+    public function validate(ServerRequestInterface $request, string $sharedSecret)
+    {
+        $hmac = $request->getHeaderLine('X-Shopify-Hmac-Sha256');
+
+        if (empty($hmac)) {
+            throw new Exception\InvalidRequestException('Incoming Shopify webhook could not be validated');
+        }
+
+        $data           = new PhpInputStream();
```
Comment (danizord) [inline src/Validator/WebhookValidator.php]: Why not `$request->getBody()`? This way you're relying on global state

File: tests/OAuth/AuthorizerTest.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+class AuthorizerTest extends \PHPUnit_Framework_TestCase
+{
+    public function shopAuthorizationProvider()
+    {
+        return [
+            ['mystore'],
+            ['mystore.myshopify.com']
+        ];
+    }
+
```
Comment (danizord) [inline tests/OAuth/AuthorizerTest.php]: Missing test for `exchangeCodeForToken()`

File: src/Validator/WebhookValidator.php (hunk 4 above)
Comment (bakura10) [inline src/Validator/WebhookValidator.php]: Does it do the same thing? I must have done this for a reason... but I don't remember :p

**Comment by bakura10** [comment]
Done @danizord, let me know if you feel anything else :).

File: src/Validator/RequestValidator.php (hunk 3 above)
Comment (danizord) [inline src/Validator/RequestValidator.php]: Benefit is that we can use this value object to hold shop domain everywhere across the apps/microservices

Comment (bakura10) [inline src/Validator/RequestValidator.php]: Makes sense. Let me add this :).

File: src/Model/ShopDomain.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    private $shopDomain;
+
+    /**
+     * @param string $shopDomain
+     */
+    public function __construct(string $shopDomain)
+    {
+        if (!preg_match('/^[a-zA-Z0-9.-]*(myshopify.com)$/', $shopDomain)) {
+            throw new Exception\RuntimeException('Shop domain is invalid');
```
Comment (danizord) [inline src/Model/ShopDomain.php]: InvalidArgument

File: src/OAuth/AuthorizationResponse.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+namespace ZfrShopify\OAuth;
+
+use Zend\Diactoros\Response\RedirectResponse;
+
+/**
+ * Create a redirection response to Shopify
+ *
+ * @author Michaël Gallego
+ */
+class AuthorizationResponse extends RedirectResponse
```
Comment (danizord) [inline src/OAuth/AuthorizationResponse.php]: `AuthorizationRedirectResponse`?

File: src/Validator/WebhookValidator.php [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+     */
+    public function validate(ServerRequestInterface $request, string $sharedSecret)
+    {
+        $hmac = $request->getHeaderLine('X-Shopify-Hmac-Sha256');
+
+        if (empty($hmac)) {
+            throw new Exception\InvalidRequestException('Incoming Shopify webhook could not be validated');
+        }
+
+        $calculatedHmac = base64_encode(hash_hmac('sha256', $request->getBody()->getContents(), $sharedSecret, true));
```
Comment (danizord) [inline src/Validator/WebhookValidator.php]: `getContents()` returns the remaining contents depending on pointer position, you can get the whole body by casting the stream to string: `(string) $request->getBody()`

**Pull request merged**